### PR TITLE
Bert ensemble

### PR DESCRIPTION
**Pull request by XiaodongWuFetch**
Finetune on IMDB dataset
Bunch of tiny changes to the whole library

### CODE REVIEW

File: libs/math/include/math/tensor.hpp [hunk 1]
```diff
@@ -1223,19 +1223,31 @@ Tensor<T, C> &Tensor<T, C>::operator=(TensorSlice const &slice)
...
 bool Tensor<T, C>::Resize(SizeVector const &shape, bool copy)
 {
-  Tensor old_tensor = *this;
+  // if the shape is exactly the same and a copy of value is required, dont do anything
+  if ((this->shape() == shape) && copy)
+  {
+    return true;
+  }
+
+  // this line does a shallow copy for speedy initializion of a tensor
```
Comment (baykaner) [inline libs/math/include/math/tensor.hpp]: pedantic point:
don't say "this line does a " in a comment - we know you're commenting about this line

File: libs/network/examples/CMakeLists.txt [hunk 2]
```diff
@@ -22,7 +22,8 @@ if (FETCH_ENABLE_EXAMPLES)
   target_link_libraries(example-network-subscribe PRIVATE fetch-network)
   target_include_directories(example-network-subscribe PRIVATE subscribe)
 
-  add_executable(example-network-listener ../../ml/include/ml/ops/activations/gelu.hpp
+  add_executable(example-network-listener ../../ml/examples/bert_for_finetune/main.cpp
```
Comment (baykaner) [inline libs/network/examples/CMakeLists.txt]: this looks extremely wrong.
what does gelu and bert have to do with the networking code?

File: libs/ml/include/ml/ops/weights.hpp [hunk 3]
```diff
@@ -215,6 +215,14 @@ class Weights : public fetch::ml::ops::PlaceHolder<T>, public Trainable<T>
       }
       break;
     }
+    case WeightsInitialisation::ONES:
+    {
+      for (std::uint64_t j = 0; j < array.data().size(); ++j)
```
Comment (baykaner) [inline libs/ml/include/ml/ops/weights.hpp]: SizeType not uint64_t

File: libs/ml/include/ml/ops/weights.hpp [hunk 4]
```diff
@@ -215,6 +215,14 @@ class Weights : public fetch::ml::ops::PlaceHolder<T>, public Trainable<T>
       }
       break;
     }
+    case WeightsInitialisation::ONES:
```
Comment (baykaner) [inline libs/ml/include/ml/ops/weights.hpp:215]: this looks very weird.
both zero and ones case should be implemented like this:

tensor.Fill(0);
tensor.Fill(1);

File: libs/ml/include/ml/layers/self_attention_encoder.hpp [hunk 5]
```diff
@@ -163,7 +163,7 @@ class SelfAttentionEncoder : public SubGraph<T>
     std::string ff_first_layer =
         this->template AddNode<fetch::ml::layers::FullyConnected<TensorType>>(
             name + "_Feedforward_No_1", {input}, static_cast<SizeType>(model_dim_),
-            static_cast<SizeType>(ff_dim_), ActivationType::RELU, RegType::NONE,
```
Comment (baykaner) [inline libs/ml/include/ml/layers/self_attention_encoder.hpp:166]: is this generically correct? or specific to BERT?
seems like this should be an input parameter if its specific to BERT

File: libs/math/tests/math/all_tests/tensor/tensor_operations.cpp [hunk 6]
```diff
@@ -273,8 +273,8 @@ TYPED_TEST(TensorOperationsTest, range_index_slices_test)
   using SizeType  = typename ArrayType::SizeType;
 
   ArrayType t1({3, 5, 2});
-  auto      s1 = t1.Slice(std::make_pair<SizeType, SizeType>(0, 3), 1);
-  assert(s1.Copy().shape() == std::vector<SizeType>({3, 3, 2}));
+  auto      s1 = t1.Slice(std::make_pair<SizeType, SizeType>(2, 4), 1);
```
Comment (baykaner) [inline libs/math/tests/math/all_tests/tensor/tensor_operations.cpp]: was this test broken?

File: libs/ml/examples/bert_for_finetune/main.cpp [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+    test_labels.Set(static_cast<SizeType>(1), static_cast<SizeType>(2) * i + 1,
+                    static_cast<DataType>(1));
+  }
+  std::cout << "finish preparing train test data" << std::endl;
+
+  return {std::make_pair(final_train_data, train_labels),
+          std::make_pair(final_test_data, test_labels)};
+}
+
+std::vector<TensorType> LoadIMDBFinetuneData(std::string const &file_path)
```
Comment (baykaner) [inline libs/ml/examples/bert_for_finetune/main.cpp:182]: move some of these into utilities.hpp

**Review by baykaner — CHANGES_REQUESTED** [review]

File: libs/ml/include/ml/layers/self_attention_encoder.hpp (hunk 5 above)
Comment (XiaodongWuFetch) [reply libs/ml/include/ml/layers/self_attention_encoder.hpp:166]: you are right

File: libs/ml/tests/ml/serializers/ml_types.cpp [hunk 8]
```diff
@@ -143,8 +143,8 @@ TYPED_TEST(SerializersTestNoInt, serialize_graph_saveable_params)
   g->SetInput("Input", data.Transpose());
   g2->SetInput("Input", data.Transpose());
 
-  TensorType prediction  = g->ForwardPropagate(output);
-  TensorType prediction2 = g2->ForwardPropagate(output);
+  TensorType prediction  = g->ForwardPropagate(output).Copy();
```
Comment (ee51kb) [inline libs/ml/tests/ml/serializers/ml_types.cpp:146]: replace with a call to Evaluate

**Review by ee51kb — APPROVED** [review]

**Review by baykaner — APPROVED** [review]

**Pull request merged**